Failure on creating pending segments table in MySQL or MariaDB metadata storage

I experienced following exception when I started Druid Overlord.
```
2016-01-20T11:32:40,984 INFO [main] com.metamx.common.lifecycle.Lifecycle$AnnotationBasedHandler - Invoking start method[public void io.druid.metadata.IndexerSQLMetadataStorageCoordinator.start()] on object[io.druid.metadata.IndexerSQLMetadataStorageCoordinator@2d7a9786].
2016-01-20T11:32:40,986 INFO [main] io.druid.metadata.SQLMetadataConnector - Creating table[druid_pendingSegments]
2016-01-20T11:32:40,997 WARN [main] com.metamx.common.RetryUtils - Failed on try 1, retrying in 1,927ms.
org.skife.jdbi.v2.exceptions.CallbackFailedException: org.skife.jdbi.v2.exceptions.UnableToExecuteStatementException: java.sql.BatchUpdateException: Specified key was too long; max key length is 1000 bytes [statement:"null", located:"null", rewritten:"null", arguments:null]
        at org.skife.jdbi.v2.DBI.withHandle(DBI.java:284) ~[jdbi-2.63.1.jar:2.63.1]
        at io.druid.metadata.SQLMetadataConnector$2.call(SQLMetadataConnector.java:110) ~[druid-server-0.9.0-SNAPSHOT.jar:0.9.0-SNAPSHOT]
        at com.metamx.common.RetryUtils.retry(RetryUtils.java:38) [java-util-0.27.6.jar:?]
        at io.druid.metadata.SQLMetadataConnector.retryWithHandle(SQLMetadataConnector.java:115) [druid-server-0.9.0-SNAPSHOT.jar:0.9.0-SNAPSHOT]
        at io.druid.metadata.SQLMetadataConnector.createTable(SQLMetadataConnector.java:160) [druid-server-0.9.0-SNAPSHOT.jar:0.9.0-SNAPSHOT]
        at io.druid.metadata.SQLMetadataConnector.createPendingSegmentsTable(SQLMetadataConnector.java:188) [druid-server-0.9.0-SNAPSHOT.jar:0.9.0-SNAPSHOT]
        at io.druid.metadata.SQLMetadataConnector.createPendingSegmentsTable(SQLMetadataConnector.java:388) [druid-server-0.9.0-SNAPSHOT.jar:0.9.0-SNAPSHOT]
        at io.druid.metadata.IndexerSQLMetadataStorageCoordinator.start(IndexerSQLMetadataStorageCoordinator.java:91) [druid-server-0.9.0-SNAPSHOT.jar:0.9.0-SNAPSHOT]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_65]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_65]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_65]
        at java.lang.reflect.Method.invoke(Method.java:497) ~[?:1.8.0_65]
        at com.metamx.common.lifecycle.Lifecycle$AnnotationBasedHandler.start(Lifecycle.java:350) [java-util-0.27.6.jar:?]
        at com.metamx.common.lifecycle.Lifecycle.start(Lifecycle.java:259) [java-util-0.27.6.jar:?]
        at io.druid.guice.LifecycleModule$2.start(LifecycleModule.java:155) [druid-api-0.3.15.jar:0.9.0-SNAPSHOT]
        at io.druid.cli.GuiceRunnable.initLifecycle(GuiceRunnable.java:73) [druid-services-0.9.0-SNAPSHOT.jar:0.9.0-SNAPSHOT]
        at io.druid.cli.ServerRunnable.run(ServerRunnable.java:40) [druid-services-0.9.0-SNAPSHOT.jar:0.9.0-SNAPSHOT]
        at io.druid.cli.Main.main(Main.java:107) [druid-services-0.9.0-SNAPSHOT.jar:0.9.0-SNAPSHOT]
Caused by: org.skife.jdbi.v2.exceptions.UnableToExecuteStatementException: java.sql.BatchUpdateException: Specified key was too long; max key length is 1000 bytes [statement:"null", located:"null", rewritten:"null", arguments:null]
        at org.skife.jdbi.v2.Batch.execute(Batch.java:131) ~[jdbi-2.63.1.jar:2.63.1]
        at io.druid.metadata.SQLMetadataConnector$4.withHandle(SQLMetadataConnector.java:172) ~[druid-server-0.9.0-SNAPSHOT.jar:0.9.0-SNAPSHOT]
        at io.druid.metadata.SQLMetadataConnector$4.withHandle(SQLMetadataConnector.java:162) ~[druid-server-0.9.0-SNAPSHOT.jar:0.9.0-SNAPSHOT]
        at org.skife.jdbi.v2.DBI.withHandle(DBI.java:281) ~[jdbi-2.63.1.jar:2.63.1]
        ... 17 more
Caused by: java.sql.BatchUpdateException: Specified key was too long; max key length is 1000 bytes
        at com.mysql.jdbc.StatementImpl.executeBatch(StatementImpl.java:1110) ~[?:?]
        at org.apache.commons.dbcp2.DelegatingStatement.executeBatch(DelegatingStatement.java:345) ~[commons-dbcp2-2.0.1.jar:2.0.1]
        at org.apache.commons.dbcp2.DelegatingStatement.executeBatch(DelegatingStatement.java:345) ~[commons-dbcp2-2.0.1.jar:2.0.1]
        at org.skife.jdbi.v2.Batch.execute(Batch.java:121) ~[jdbi-2.63.1.jar:2.63.1]
        at io.druid.metadata.SQLMetadataConnector$4.withHandle(SQLMetadataConnector.java:172) ~[druid-server-0.9.0-SNAPSHOT.jar:0.9.0-SNAPSHOT]
        at io.druid.metadata.SQLMetadataConnector$4.withHandle(SQLMetadataConnector.java:162) ~[druid-server-0.9.0-SNAPSHOT.jar:0.9.0-SNAPSHOT]
        at org.skife.jdbi.v2.DBI.withHandle(DBI.java:281) ~[jdbi-2.63.1.jar:2.63.1]
        ... 17 more
```

--- a/server/src/main/java/io/druid/metadata/IndexerSQLMetadataStorageCoordinator.java
+++ b/server/src/main/java/io/druid/metadata/IndexerSQLMetadataStorageCoordinator.java
@@ -29,8 +29,11 @@ import com.google.common.collect.Iterables;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Ordering;
 import com.google.common.collect.Sets;
+import com.google.common.hash.Hashing;
+import com.google.common.io.BaseEncoding;
 import com.google.inject.Inject;
 import com.metamx.common.IAE;
+import com.metamx.common.StringUtils;
 import com.metamx.common.lifecycle.LifecycleStart;
 import com.metamx.common.logger.Logger;
 import io.druid.indexing.overlord.IndexerMetadataStorageCoordinator;
@@ -447,10 +450,23 @@ public class IndexerSQLMetadataStorageCoordinator implements IndexerMetadataStor
             // Avoiding ON DUPLICATE KEY since it's not portable.
             // Avoiding try/catch since it may cause inadvertent transaction-splitting.
 
+            // UNIQUE key for the row, ensuring sequences do not fork in two directions.
+            // Using a single column instead of (sequence_name, sequence_prev_id) as some MySQL storage engines
+            // have difficulty with large unique keys (see https://github.com/druid-io/druid/issues/2319)
+            final String sequenceNamePrevIdSha1 = BaseEncoding.base16().encode(
+                Hashing.sha1()
+                       .newHasher()
+                       .putBytes(StringUtils.toUtf8(sequenceName))
+                       .putByte((byte) 0xff)
+                       .putBytes(StringUtils.toUtf8(previousSegmentIdNotNull))
+                       .hash()
+                       .asBytes()
+            );
+
             handle.createStatement(
                 String.format(
-                    "INSERT INTO %s (id, dataSource, created_date, start, \"end\", sequence_name, sequence_prev_id, payload) "
-                    + "VALUES (:id, :dataSource, :created_date, :start, :end, :sequence_name, :sequence_prev_id, :payload)",
+                    "INSERT INTO %s (id, dataSource, created_date, start, \"end\", sequence_name, sequence_prev_id, sequence_name_prev_id_sha1, payload) "
+                    + "VALUES (:id, :dataSource, :created_date, :start, :end, :sequence_name, :sequence_prev_id, :sequence_name_prev_id_sha1, :payload)",
                     dbTables.getPendingSegmentsTable()
                 )
             )
@@ -461,6 +477,7 @@ public class IndexerSQLMetadataStorageCoordinator implements IndexerMetadataStor
                   .bind("end", interval.getEnd().toString())
                   .bind("sequence_name", sequenceName)
                   .bind("sequence_prev_id", previousSegmentIdNotNull)
+                  .bind("sequence_name_prev_id_sha1", sequenceNamePrevIdSha1)
                   .bind("payload", jsonMapper.writeValueAsBytes(newIdentifier))
                   .execute();
 

--- a/server/src/main/java/io/druid/metadata/SQLMetadataConnector.java
+++ b/server/src/main/java/io/druid/metadata/SQLMetadataConnector.java
@@ -197,9 +197,10 @@ public abstract class SQLMetadataConnector implements MetadataStorageConnector
                 + "  \"end\" VARCHAR(255) NOT NULL,\n"
                 + "  sequence_name VARCHAR(255) NOT NULL,\n"
                 + "  sequence_prev_id VARCHAR(255) NOT NULL,\n"
+                + "  sequence_name_prev_id_sha1 VARCHAR(255) NOT NULL,\n"
                 + "  payload %2$s NOT NULL,\n"
                 + "  PRIMARY KEY (id),\n"
-                + "  UNIQUE (sequence_name, sequence_prev_id)\n"
+                + "  UNIQUE (sequence_name_prev_id_sha1)\n"
                 + ")",
                 tableName, getPayloadType()
             )

